Fix bug in `local_elemwise_alloc` and add tests

ricardoV94 opened the pull request:
Rewrite would return a new graph identical to itself, resulting in an endless rewrite cycle

```python
import aesara
import aesara.tensor as at

x = at.matrix("x")
y = at.matrix("y")
z = at.mul(at.alloc(x, 1, 5), y)

with aesara.config.change_flags(optimizer_verbose=True):
    f = aesara.function([x, y], z, mode="FAST_RUN")
```
```
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
optimizer: rewrite local_elemwise_alloc replaces Elemwise{mul,no_inplace}.0 with Elemwise{mul,no_inplace}.0
...
```

~There are still flaws in the rewrite logic, as indicated by the new triggered assertion~

First showed up in https://github.com/pymc-devs/pymc/pull/5369

TODO:
- [x] Rewrite docstrings to make clear what is going on in this rewrite
- [ ] Make sure tests cover all rewrite cases
- [ ] Consider removing the special `Dimshuffle` of `Alloc` logic

ricardoV94 commented on aesara/tensor/basic_opt.py:
```diff
@@ -1645,6 +1645,10 @@ def dimshuffled_alloc(i):
             new_i.append(i)
```
Problem is that we hit this `else` bucket which should only apply to the inputs that are not relevant for the rewrite...

ricardoV94 commented on aesara/tensor/basic_opt.py:
```diff
@@ -1610,7 +1610,18 @@ def dimshuffled_alloc(i):
-            new_i.append(i.owner.inputs[0])
+            alloc_input = i.owner.inputs[0]
+            if alloc_input.ndim != i.owner.inputs[0].ndim:
+                # The `Alloc` can add dimension to the value
```
This looks like a very odd thing. Should we just intercept `Alloc` calls that add new dimensions and return `dimshuffles` instead?

brandonwillard commented on aesara/tensor/basic_opt.py:
```diff
@@ -1626,15 +1630,15 @@ def dimshuffled_alloc(i):
-                # The `Alloc` can add dimension to the value
-                # We add a `DimShuffle` to add them.
-                # We let later optimization merge the multiple `DimShuffle`
+                # The `Alloc` can add dimensions to the value.
+                # We replace those cases with a `DimShuffle` here.
+                # We let later optimizations merge the nested `DimShuffle`s
```
Can you mention why this is important to do (e.g. to avoid non-termination)?

brandonwillard commented on aesara/tensor/basic_opt.py:
```diff
@@ -1645,6 +1649,10 @@ def dimshuffled_alloc(i):
             new_i.append(i)
     new_i[assert_op_idx] = assert_op_in
 
+    # If this assert is triggered, it means we are recreating an equivalent graph
+    # which would result in a cyclical merge optimization
+    assert not all(new is old for new, old in zip(new_i, node.inputs))
```
Are we not sure that the changes above will necessarily prevent this from occurring?  If not, perhaps we need to put the logic of this statement somewhere above.

ricardoV94 replied:
I kept finding more corner cases... Hopefully there are no others by this point.

brandonwillard replied:
Couldn't the be used as a `return` condition as well?

ricardoV94 replied:
It could... but still means we have some flawed logic above, where the "feasibility" of the rewrite is supposed to take place. What do you think is best? This test has a couple of other asserts in place, and I have no idea of how unconventional that is.

brandonwillard reviewed (dismissed):
Can you add a MWE that reproduces the issue in the description (and explicitly shows the problematic graphs)?

ricardoV94 commented on aesara/tensor/basic_opt.py:
```diff
@@ -1610,7 +1604,17 @@ def dimshuffled_alloc(i):
-            new_i.append(i.owner.inputs[0])
+            alloc_input = i.owner.inputs[0]
+            if alloc_input.ndim != i.ndim:
+                # The `Alloc` can add dimensions to the value.
+                # We replace those cases with a `DimShuffle` here.
```
This matters in some new edge cases I found while fixing the original issues. Basically it led to the same type of bug, we were returning a new "graph" that was exactly like the original one without realizing it, because we were not addressing all possible cases for an `Alloc` inside the `Elemwise`

brandonwillard replied:
That set of comments needs to mention how this avoids producing the same graph; otherwise, it won't help future readers very much.

ricardoV94 replied:
The test is about removing `Alloc`s so this shouldn't sound too surprising. It is only surprising in the light of the bug.

brandonwillard commented on tests/tensor/test_basic_opt.py:
```diff
@@ -3505,3 +3505,28 @@ def test_Shape_i_canonicalize():
+@pytest.mark.parametrize(
+    "expr",
+    [
+        (lambda x, y: at.mul(at.alloc(x, 15, 1), y)),
+        (lambda x, y: at.mul(at.alloc(x, 15, 2), y)),
+        (lambda x, y: at.mul(at.alloc(x, 15, 1), at.alloc(y, 15, 1))),
+        (lambda x, y: at.mul(at.alloc(x, 15, 2), at.alloc(y, 15, 2))),
+        (lambda x, y: at.mul(at.alloc(x, 15, 2).dimshuffle(1, 0), y)),
+        (lambda x, y: at.mul(at.alloc(x, 1, 15, 2), y)),
+        (lambda x, y: at.mul(at.alloc(x, 1, 15, 2).dimshuffle(0, 2, 1), y)),
```
I'm not able to reproduce the issue on `main` with any of these cases.  Can you produce a debug print of one such graph?

brandonwillard replied:
Nevermind, got it now.

ricardoV94 commented:
> Can you add a MWE that reproduces the issue in the description?

Done

brandonwillard commented:
I'm starting to see that `local_elemwise_alloc` is one of those old rewrites that needs to be completely rewritten.

ricardoV94 commented:
I don't quite get how this rewrite is supposed to work now that I think about it. If we start with `Elemwise(Alloc1(x, shape), Alloc2(y, shape))` and then remove `Alloc1` because we have `Alloc2` -> `Elemwise(x, Alloc2(shape))`, how do we prevent `Alloc2` to be removed in the next iteration? If we do that, we lost all shape information no? In other words, these new tests should actually fail:

```python
(lambda x, y: at.mul(at.alloc(x, 15, 1), at.alloc(y, 15, 1))),
(lambda x, y: at.mul(at.alloc(x, 15, 2), at.alloc(y, 15, 2))),
```

brandonwillard commented:
> I don't quite get how this rewrite is supposed to work now that I think about it. If we start with `Elemwise(Alloc1(x, shape), Alloc2(y, shape))` and then remove `Alloc1` because we have `Alloc2` -> `Elemwise(x, Alloc2(y, shape))`, how do we prevent `Alloc2` from being removed in the next iteration? If we do that, we lost all shape information, no? In other words, these new tests should actually fail:
> 
> ```python
> (lambda x, y: at.mul(at.alloc(x, 15, 1), at.alloc(y, 15, 1))),
> (lambda x, y: at.mul(at.alloc(x, 15, 2), at.alloc(y, 15, 2))),
> ```
> 
> Nevermind, that's what the `asssert_op_in` does

Yeah, the docstring and logic are both unnecessarily confusing and convoluted.  That&mdash;combined with a lack of quality (e.g. direct) unit tests&mdash;almost necessarily give rise to these kinds of issues.

This is yet another example of why we need to move toward formalizing our rewrites and their descriptions.  In this case, the docstring uses its own notation (e.g. `y.TensorType(BROADCAST CONDITION)`), and that only serves to make things _more_ confusing.

ricardoV94 commented:
I cleaned a bit the test logic in a separate commit. The two test conditions with the `dimshuffle` fail (now that I removed `"specialize"` from the `include`, because those `Alloc` remains in the `Assert` that is added around `y`. 

This is not necessarily a failure, but suggests the original rewrite was not quite able to deliver what it meant to?

brandonwillard commented:
> This is not necessarily a failure, but suggests the original rewrite was not quite able to deliver what it meant to?

Wouldn't surprise me, but, while we're fixing this, let's at least make our assumptions/goals very clear in the docstring.

Also, we need to do something about that `DimShuffle` complexity.  I think it's nearly doubling the logic in this rewrite and it doesn't even need to be there.

Is there a `DimShuffle` lift/removal that can be applied to `Alloc`s independently.  For example, `DimShuffle{"x", ...}(Alloc(y, s1, ...)) -> Alloc(y, 1, s1, ...)`?

ricardoV94 commented:
> > This is not necessarily a failure, but suggests the original rewrite was not quite able to deliver what it meant to?
> 
> Wouldn't surprise me, but, while we're fixing this, let's at least make our assumptions/goals very clear in the docstring.
> 
> Also, we need to do something about that `DimShuffle` complexity. I think it's nearly doubling the logic in this rewrite and it doesn't even need to be there.
> 
> Is there a `DimShuffle` lift/removal that can be applied to `Alloc`s independently. For example, `DimShuffle{"x", ...}(Alloc(y, s1, ...)) -> Alloc(y, 1, s1, ...)`?

Yes that could be done, but according to the comments in the docstrings that would be less efficient in the general case (where `Alloc` stay in the graph)

brandonwillard commented:
> Yes that could be done, but according to the comments in the docstrings that would be less efficient in the general case (where `Alloc` stay in the graph)

I don't see how that could be a real problem (for at least the case of adding broadcastable dimensions to an `Alloc`), and, if it actually was realistically more efficient, wouldn't we want an independent "specialization" rewrite that changes _all_ such `Alloc`s into `DimShuffle`s?  It wouldn't make any sense to limit that to the context of this rewrite only.

Here's a quick and dirty check:
```python
def alloc(v, shape):
    res = np.empty(shape)
    res[...] = v
    return res


%timeit alloc(2, (1000, 1000))
258 µs ± 8.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit alloc(2, (3000, 2000))
5.71 ms ± 87.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit alloc(2, (1, 1, 1, 1, 1000, 1, 1, 1000, 1, 1, 1))
280 µs ± 22.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit alloc(2, (1, 1, 1, 1, 3000, 1, 1, 2000, 1, 1, 1))
5.8 ms ± 134 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

The old codebase has some confusing "efficiency"-based comments like that&mdash;ones that I'm not sure were ever necessarily true or relevant. 

Regardless of the practical truth of these kinds of statements, the choices that appear to have resulted from them clearly led to effectively contradictory concerns and approaches.

brandonwillard commented:
The docstring of `local_canonicalize_alloc` mentions this `Alloc`-to-`DimShuffle` topic and says https://github.com/Theano/Theano/issues/4072 provides reasons for why it's needed.  It don't see anything about performance&mdash;or any real justifications other than that it might've helped resolve a different problem by generating graphs that some logic could already handle.

In other words, the rewrites/canonical form for `Alloc` is a mess.
